tests/docs/test_supported_confs.py

```diff
@@ -0,0 +1,493 @@
+                return probs[0][0]
+
+            # compute the gradient of the scalar cost function instead
+            # of the jacobian of the state directly - the latter is a
+            # separate test case
+            res = cost_fn(x)
+            res.backward()
+            return x.grad
+        else:
+            return F.jacobian(circuit, (x,))
+
+
+class TestSupportedConfs:
+    """Test that the supported configurations in the documentation
+    matches the supported configurations in the code"""
+
+    @pytest.mark.parametrize("interface", interfaces)
+    @pytest.mark.parametrize("return_type", return_types)
+    @pytest.mark.parametrize("shots", shotss)
+    def test_all_device(self, interface, return_type, shots):
+        """Test diff_method=device raises an error for all interfaces"""
+        msg = (
+            "The default.qubit device does not provide a native "
+            "method for computing the jacobian."
+        )
+
+        with pytest.raises(QuantumFunctionError, match=msg):
+            circuit = get_qnode(interface, "device", return_type, shots)
+
+    @pytest.mark.parametrize("return_type", return_types)
+    def test_none_backprop(self, return_type):
+        """Test interface=None and diff_method=backprop raises an error"""
+        msg = (
+            "Device default.qubit only supports diff_method='backprop' when "
+            "using the ['tf', 'torch', 'autograd', 'jax'] interfaces."
+        )
+        msg = re.escape(msg)
+
+        with pytest.raises(QuantumFunctionError, match=msg):
+            circuit = get_qnode(None, "backprop", return_type, None)
+
+    @pytest.mark.parametrize("diff_method", ["adjoint", "parameter-shift", "finite-diff"])
+    @pytest.mark.parametrize("return_type", return_types)
+    @pytest.mark.parametrize("shots", shotss)
+    def test_none_all(self, diff_method, return_type, shots):
+        """Test interface=None and diff_method in [adjoint, parameter-shift,
+        finite-diff] has a working forward pass"""
+        warn_msg = (
+            "Requested adjoint differentiation to be computed with finite shots. "
+            "Adjoint differentiation always calculated exactly."
+        )
+
+        if diff_method == "adjoint" and shots is not None:
+            # this warning is still raised in the forward pass
+            with pytest.warns(UserWarning, match=warn_msg):
+                circuit = get_qnode(None, diff_method, return_type, shots)
+        else:
+            circuit = get_qnode(None, diff_method, return_type, shots)
+
+    @pytest.mark.parametrize("interface", diff_interfaces)
+    @pytest.mark.parametrize(
+        "return_type",
+        [State, "DensityMatrix", Probability, Expectation, "Hermitian", "Projector", Variance],
+    )
+    def test_all_backprop_none_shots(self, interface, return_type):
+        """Test diff_method=backprop works for all interfaces when shots=None"""
+
+        # DensityMatrix doesn't work with torch at the moment
```

Comment:
Maybe the issue number could also be added here (potentially with a link to the issue from GitHub).

Reply:
Added TODO with link to issue.